SPECS/tdnf/virtual-repo-snapshot.patch

```diff
@@ -0,0 +1,906 @@
+     } while(0)
+ 
++typedef struct {
++    // frequently changed values
++    char * pszElementBuffer;
++    int nBufferLen;
++    int nInPackage;
++    int nPrintPackage;
++    int nTimeFound;
++
++    // managed values
++    int nBufferMaxLen;
++    int nDepth;
++    int nPrevElement; // enum 0 -> start, 1 -> data, 2 -> end
++
++    //set and forget on creation
++    time_t nSearchTime;
++    FILE * pbOutfile;
++} TDNFFilterData;
++
++#define TDNF_MAX_FILTER_INPUT_THRESHOLD 10000000
++
++#define BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError) \
++    do {                                                           \
++        if (dwError)                                               \
++        {                                                          \
++            goto error;                                            \
++        }                                                          \
++    } while(0)
++
+ #endif /* __SOLV_DEFINES_H__ */
+diff --git a/solv/includes.h b/solv/includes.h
+index 2ab0c5c..1e39ecb 100644
+--- a/solv/includes.h
++++ b/solv/includes.h
+@@ -10,7 +10,6 @@
+ #include <unistd.h>
+ #include <stdbool.h>
+ #include <sys/types.h>
+-#include <sys/stat.h>
+ #include <dirent.h>
+ 
+ // libsolv
+@@ -44,4 +43,9 @@
+ #include "../history/history.h"
+ #include "prototypes.h"
+ 
++#include <expat.h>
++#include <stdlib.h>
++#include <time.h>
++#include <string.h>
++
+ #endif /* __SOLV_INCLUDES_H__ */
+diff --git a/solv/prototypes.h b/solv/prototypes.h
+index 2633b5e..0f5df8d 100644
+--- a/solv/prototypes.h
++++ b/solv/prototypes.h
+@@ -517,7 +517,8 @@ SolvReadYumRepo(
+     const char *pszPrimary,
+     const char *pszFilelists,
+     const char *pszUpdateinfo,
+-    const char *pszOther
++    const char *pszOther,
++    char * pszRepoEpochLimit
+     );
+ 
+ uint32_t
+diff --git a/solv/tdnfrepo.c b/solv/tdnfrepo.c
+index c27b907..aef6c61 100644
+--- a/solv/tdnfrepo.c
++++ b/solv/tdnfrepo.c
+@@ -9,6 +9,522 @@
+ 
+ #include "includes.h"
+ 
++// #### XML FILTER CODE ####
++void checkAndResizeBuffer(char ** pszCharBuffer, int * nBufferMaxLen, int nLengthToAdd) {
++    uint32_t dwError = 0;
++    int nBufferContentLen = strlen(*pszCharBuffer);
++    while (nBufferContentLen + nLengthToAdd >= *nBufferMaxLen)
++    {
++        *nBufferMaxLen *= 2;
++        char * pszTempCharBuffer = NULL;
++        pszTempCharBuffer = realloc(*pszCharBuffer, *nBufferMaxLen);
++        if (!pszTempCharBuffer)
++        {
++            dwError = ERROR_TDNF_TIME_FILTER_MEMORY;
++            BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError);
++        }
++        //set expanded char buffer 
++        *pszCharBuffer = pszTempCharBuffer;
++    }
++cleanup:
++    return;
++error:
++    goto cleanup;
++}
++
++char * descEscapeCharLinter(char * desc) {
++    uint32_t dwError = 0;
++    char * amp = "&amp;";
++    char * gt = "&gt;";
++    char * lt = "&lt;";
++
++    int nDescLen = strlen(desc) * sizeof(char) + 1;
```

Comment:
not sure if they are using UTF8 here.  If so, +1 won't work.  Also, why +1?

Reply:
Good catch, I have moved the +1 to happen prior to the multiplication of size.